### PR TITLE
Fail CI if diff not empty during tests

### PR DESCRIPTION
**Pull request by leudz**
## 🌟 What is the purpose of this PR?

Making sure no generated file were forgotten when opening a PR.
This can happen with `Cargo.lock` for example.

## 🔍 What does this change?

- Add a check to make sure that the `git diff` is empty after tests.

### 📜 Does this require a change to the docs?

No

## ❓ How to test this?

In theory push a commit with an outdated `Cargo.lock`.
You can see it on this commit https://github.com/hashintel/hash/runs/6120802375?check_suite_focus=true.
As opposed to the same code but with the `Cargo.lock` https://github.com/hashintel/hash/pull/498/checks?sha=3516ca3216780945ad2c8140b4b18ff2af4f8a4d.

### CODE REVIEW

**Comment by TimDiekmann** [comment]
You may just throw in
```bash
git --no-pager diff --exit-code --color
```
after the jobs. This is printing a diff view, returns non-zero if it's not empty and it's colored.

**Comment by TimDiekmann** [comment]
Maybe something like that:
```yaml
      - name: Ensure empty git diff
        run: git --no-pager diff --exit-code --color
```
If you come up with a better name I'd appreciate that 😉

**Comment by leudz** [comment]
@TimDiekmann The command worked great! Definitely simpler than what I found :facepalm: 
I wanted to also put it before the tests because I expected tests to take a long time and wanted faster feedback on that. But some tests are very fast so it doesn't matter.

For the name maybe something along the lines of "Ensure no generated files changed". I don't think the diff will catch anything other than files generated by compilation.

Also the test currently fails as you can see here: https://github.com/hashintel/hash/runs/6120820355?check_suite_focus=true.
I think it comes from this https://github.com/hashintel/hash/blob/90ea01e0a22e83b2c5e6d907e7f457cc8c5a575f/.github/workflows/rust.yml#L206.
Do we want to store this change and add it back before the diff?

**Comment by TimDiekmann** [comment]
> I wanted to also put it before the tests because I expected tests to take a long time and wanted faster feedback on that. But some tests are very fast so it doesn't matter.

I wouldn't like to anyway because if we have such a test it makes sense to also verify, that the tests do not generate any files which are not ignored.

> For the name maybe something along the lines of "Ensure no generated files changed". I don't think the diff will catch anything other than files generated by compilation.

I agree that **most likely**, this will only happen with generated files but it's not limited to them.


> Do we want to store this change and add it back before the diff?

That's unfortunate but, yes, let's do this.

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -247,6 +247,10 @@ jobs:
       - name: Run tests
         run: cargo test --manifest-path ${{ matrix.directory }}/Cargo.toml --no-fail-fast ${{ matrix.flags }}
 
+      # Non empty diff can be a Cargo.lock that wasn't pushed.
```
Comment (TimDiekmann) [inline .github/workflows/rust.yml:250]:
```suggestion
      # Non-empty diff could be a Cargo.lock that wasn't committed.
```

**Comment by leudz** [comment]
> I wouldn't like to anyway because if we have such a test it makes sense to also verify, that the tests do not generate any files which are not ignored.

I had the check twice, one before and one after tests.

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -202,8 +202,15 @@ jobs:
...
-      - name: Remove default members
-        run: sed '/default-members/d' -i ${{ matrix.directory }}/Cargo.toml
+      # Store `Cargo.toml` before modifying it to not affect the diff.
+      - name: Remove and store default members
+        id: default_members
+        run: |
+          cargo_toml=$(< ${{ matrix.directory }}/Cargo.toml)
+          echo "ORIGINAL_CARGO_TOML<<EOF" >> $GITHUB_ENV
+          echo "$cargo_toml" >> $GITHUB_ENV
+          echo "EOF" >> $GITHUB_ENV
```
Comment (TimDiekmann) [inline .github/workflows/rust.yml:212]: Instead of dancing around with the environment variable, can't we simple do this?
```bash
cp "${{ matrix.directory }}/Cargo.toml" "${{ matrix.directory }}/Cargo.toml.orig"
```
and later
```bash
mv "${{ matrix.directory }}/Cargo.toml.orig" "${{ matrix.directory }}/Cargo.toml"
```

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -202,8 +202,11 @@ jobs:
       - name: Checkout source code
         uses: actions/checkout@v2
 
-      - name: Remove default members
-        run: sed '/default-members/d' -i ${{ matrix.directory }}/Cargo.toml
+      # Store `Cargo.toml` before modifying it to not affect the diff.
+      - name: Remove and store default members
+        run: |
+          cp "${{ matrix.directory }}/Cargo.toml" "${{ matrix.directory }}/Cargo.toml.orig"
```
Comment (TimDiekmann) [inline .github/workflows/rust.yml:208]:
```suggestion
      - name: Remove and store default members
        run: |
          # Backup `Config.toml` to restore it later so that `git diff` does not report any changes
          cp "${{ matrix.directory }}/Cargo.toml" "${{ matrix.directory }}/Cargo.toml.orig"
```

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -247,6 +250,12 @@ jobs:
       - name: Run tests
         run: cargo test --manifest-path ${{ matrix.directory }}/Cargo.toml --no-fail-fast ${{ matrix.flags }}
 
+      # Non-empty diff could be a Cargo.lock that wasn't committed.
+      - name: Ensure empty git diff
+        run: |
+          mv "${{ matrix.directory }}/Cargo.toml.orig" "${{ matrix.directory }}/Cargo.toml"
```
Comment (TimDiekmann) [inline .github/workflows/rust.yml:256]:
```suggestion
          mv -v "${{ matrix.directory }}/Cargo.toml.orig" "${{ matrix.directory }}/Cargo.toml"
```

**Review by TimDiekmann — APPROVED** [review]

File: .github/workflows/rust.yml (hunk 2 above)
Comment (leudz) [reply .github/workflows/rust.yml]: Yeah the env var started simple and ended up sooo complicated...

Comment (TimDiekmann) [reply .github/workflows/rust.yml]: I actually liked the idea of the environment variable, I wasn't aware of `$GITHUB_ENV`, this could be useful in other places as well. Also using an output is probably the cleanest way but very verbose and hard to follow. This solution is probably the simplest **and most robust** one.
